File: oop/05-interfaces/csharp/BankAccount.cs
/*
 * Demonstrates: Interfaces - Multiple Interface Implementation
 * Concept: BankAccount implements both ITransactable and IAuditable
 */

namespace FinanceDomain.Interfaces;

public class BankAccount : ITransactable, IAuditable
{
    private decimal _balance;
    private readonly List<AuditEntry> _auditLog = [];

    public string AccountNumber { get; }
    public string AccountHolder { get; }
    public decimal Balance => _balance;

    public BankAccount(string accountNumber, string accountHolder, decimal openingBalance)
    {
        AccountNumber = accountNumber;
        AccountHolder = accountHolder;
        _balance = openingBalance;
        RecordAudit("ACCOUNT_OPENED", $"Opening balance: R{openingBalance:F2}");
    }

    public void Deposit(decimal amount)
    {
        if (amount <= 0)
            throw new ArgumentException("Deposit amount must be greater than zero.");

        _balance += amount;
        RecordAudit("DEPOSIT", $"Amount: R{amount:F2} | New Balance: R{_balance:F2}");
    }

    public void Withdraw(decimal amount)
    {
        if (amount <= 0)
            throw new ArgumentException("Withdrawal amount must be greater than zero.");

        if (amount > _balance)
            throw new InvalidOperationException($"Insufficient funds. Available: R{_balance:F2}");

        _balance -= amount;
        RecordAudit("WITHDRAWAL", $"Amount: R{amount:F2} | New Balance: R{_balance:F2}");
    }

    public IReadOnlyList<AuditEntry> GetAuditLog() => _auditLog.AsReadOnly();

    public void RecordAudit(string action, string details)
    {
        _auditLog.Add(new AuditEntry(action, details));
    }
}

File: oop/05-interfaces/typescript/bankAccount.ts
/*
 * Demonstrates: Interfaces - Multiple Interface Implementation
 * Concept: BankAccount implements both ITransactable and IAuditable
 */

import { ITransactable } from "./transactable";
import { IAuditable, AuditEntry } from "./auditable";

export class BankAccount implements ITransactable, IAuditable {
    private _balance: number;
    private readonly auditLog: AuditEntry[] = [];

    public readonly accountNumber: string;
    public readonly accountHolder: string;

    constructor(accountNumber: string, accountHolder: string, openingBalance: number) {
        this.accountNumber = accountNumber;
        this.accountHolder = accountHolder;
        this._balance = openingBalance;
        this.recordAudit("ACCOUNT_OPENED", `Opening balance: R${openingBalance.toFixed(2)}`);
    }

    get balance(): number {
        return this._balance;
    }

    deposit(amount: number): void {
        if (amount <= 0) {
            throw new Error("Deposit amount must be greater than zero.");
        }
        this._balance += amount;
        this.recordAudit("DEPOSIT", `Amount: R${amount.toFixed(2)} | New Balance: R${this._balance.toFixed(2)}`);
    }

    withdraw(amount: number): void {
        if (amount <= 0) {
            throw new Error("Withdrawal amount must be greater than zero.");
        }
        if (amount > this._balance) {
            throw new Error(`Insufficient funds. Available: R${this._balance.toFixed(2)}`);
        }
        this._balance -= amount;
        this.recordAudit("WITHDRAWAL", `Amount: R${amount.toFixed(2)} | New Balance: R${this._balance.toFixed(2)}`);
    }

    getAuditLog(): readonly AuditEntry[] {
        return [...this.auditLog];
    }

    recordAudit(action: string, details: string): void {
        this.auditLog.push(new AuditEntry(action, details));
    }
}
